support/controlplane-component/status.go

```diff
@@ -0,0 +1,167 @@
+		if availableCondition != nil && availableCondition.Status == metav1.ConditionTrue {
+			unavailableDependencies.Delete(component.Name)
+		}
+	}
+
+	return sets.List(unavailableDependencies), nil
+}
+
+func (c *controlPlaneWorkload) reconcileComponentStatus(cpContext ControlPlaneContext, component *hyperv1.ControlPlaneComponent, unavailableDependencies []string, reconcilationError error) error {
+	component.Status.Resources = []hyperv1.ComponentResource{}
+	if err := assets.ForEachManifest(c.Name(), func(manifestName string) error {
+		adapter, exist := c.manifestsAdapters[manifestName]
+		if exist && adapter.predicate != nil && !adapter.predicate(cpContext) {
+			return nil
+		}
+
+		obj, gvk, err := assets.LoadManifest(c.name, manifestName)
+		if err != nil {
+			return err
+		}
+
+		component.Status.Resources = append(component.Status.Resources, hyperv1.ComponentResource{
+			Kind:  gvk.Kind,
+			Group: gvk.Group,
+			Name:  obj.GetName(),
+		})
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if len(unavailableDependencies) > 0 {
+		meta.SetStatusCondition(&component.Status.Conditions, metav1.Condition{
+			Type:    string(hyperv1.ControlPlaneComponentAvailable),
+			Status:  metav1.ConditionFalse,
+			Reason:  hyperv1.WaitingForDependenciesReason,
+			Message: fmt.Sprintf("Waiting for Dependencies: %s", strings.Join(unavailableDependencies, ", ")),
+		})
+		return nil
+	}
+
+	if reconcilationError != nil {
+		meta.SetStatusCondition(&component.Status.Conditions, metav1.Condition{
+			Type:    string(hyperv1.ControlPlaneComponentAvailable),
+			Status:  metav1.ConditionFalse,
+			Reason:  hyperv1.ReconciliationErrorReason,
+			Message: reconcilationError.Error(),
+		})
+		return nil
+	}
+
+	// set version status only if there was no reconcilationError
+	component.Status.Version = cpContext.ReleaseImageProvider.Version()
+
+	var status metav1.ConditionStatus
+	var reason, message string
+	if c.workloadType == deploymentWorkloadType {
+		status, reason, message = c.isDeploymentReady(cpContext)
+	} else {
+		status, reason, message = c.isStatefulSetReady(cpContext)
+	}
+
+	meta.SetStatusCondition(&component.Status.Conditions, metav1.Condition{
+		Type:    string(hyperv1.ControlPlaneComponentAvailable),
+		Status:  status,
+		Reason:  reason,
+		Message: message,
+	})
+	return nil
+}
+
+func (c *controlPlaneWorkload) isDeploymentReady(cpContext ControlPlaneContext) (status metav1.ConditionStatus, reason string, message string) {
+	status = metav1.ConditionFalse
+	deployment := &appsv1.Deployment{}
+	if err := cpContext.Client.Get(cpContext, client.ObjectKey{Namespace: cpContext.HCP.Namespace, Name: c.name}, deployment); err != nil {
+		if !apierrors.IsNotFound(err) {
+			reason = "Error"
+			message = err.Error()
+		} else {
+			reason = hyperv1.NotFoundReason
+			message = fmt.Sprintf("%s Deployment not found", deployment.Name)
+		}
+		return
+	}
+
+	for _, cond := range deployment.Status.Conditions {
```

Comment:
we already have a function for this
https://github.com/openshift/hypershift/blob/a0191dbda4ac75bd8ee19869d9a952aa508b3f2b/support/util/deployment.go#L15

Reply:
I'd rather use the Deployment available condition here, so we stay consistent with however k8s defines a deployment to be ready, and also to use the condition message when its false.

Reply:
My understanding, the difference between ours and the upstream kube is that we take into consideration whether the latest spec is available, not just whether the deployment is available.